libpysal/weights/util.py

```diff
@@ -211,6 +207,82 @@ def lat2W(nrows=5, ncols=5, rook=True, id_type='int', **kwargs):
     return W(w, weights, ids=ids, id_order=ids[:], **kwargs)
 
 
+def xrast2W(rasterf, rook=True, **kwargs):
+    """
+    Create a W object from rasters(xarray.DataArray)
+    
+    Parameters
+    ----------
+    raster     : xarray.DataArray
+                 raster file accessed using xarray.open_rasterio method
+    rook       : boolean
+                 type of contiguity. Default is rook. For queen, rook =False
+    **kwargs   : keyword arguments
+                 optional arguments for :class:`pysal.weights.W`
+    Returns
+    -------
+    w    : libpysal.weights.W
+           instance of spatial weights class W
+    data : pandas.Series
+           Values from `raster` as a vector of dimension (`w.n` x
+           None) aligned with `w`            
+    """
+    try:
+        import pandas as pd
+        import xarray as xr
+    except ImportError:
+        raise ImportError(
+            "pandas and xarray must be installed to use this method")
+
+    w = lat2W(*rasterf[0].shape, rook=rook, **kwargs)
+    if isinstance(rasterf, xr.DataArray):
+        rasterf_masked = rasterf.where(rasterf.data != rasterf.nodatavals[0])
+        rasterf_masked = rasterf_masked[0].data.flatten()
```

Comment:
It'd be great to at least explore the option of performing this slicing within the sparse format. Ideally, we want to avoid having to go from sparse to dense to sparse again. For performance reasons but, particularly, for memory efficiency.

Reply:
sure, i think i got it... will push the changes soon